cmd/gke-serviceaccounts-initializer/main.go

```diff
@@ -0,0 +1,223 @@
+		kubecfg := filepath.Join(os.Getenv("HOME"), ".kube", "config")
+		log.Printf("Using kubeconfig file at %s", kubecfg)
+		clusterConfig, err = clientcmd.BuildConfigFromFlags("", kubecfg)
+		if err != nil {
+			log.Printf("failed to find kubeconfig file: %+v", err)
+			log.Fatal("No authentication is available.")
+		}
+	}
+
+	clientset, err := kubernetes.NewForConfig(clusterConfig)
+	if err != nil {
+		log.Fatalf("failed to initialize kubernetes client: %+v", err)
+	}
+
+	// Watch uninitialized Pods in all namespaces.
+	restClient := clientset.CoreV1().RESTClient()
+	watchlist := cache.NewListWatchFromClient(restClient,
+		"pods", corev1.NamespaceAll, fields.Everything())
+
+	// Wrap the returned watchlist to workaround the inability to include
+	// the `IncludeUninitialized` list option when setting up watch clients.
+	includeUninitializedWatchlist := &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			options.IncludeUninitialized = true
+			return watchlist.List(options)
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			options.IncludeUninitialized = true
+			return watchlist.Watch(options)
+		},
+	}
+
+	_, controller := cache.NewInformer(includeUninitializedWatchlist,
+		&corev1.Pod{},
+		resyncPeriod,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				pod, ok := obj.(*corev1.Pod)
+				if !ok {
+					log.Fatalf("watch returned non-pod object: %T", pod)
+				}
+
+				if !needsInitialization(pod) {
+					log.Printf("skipping pod/%s", pod.GetName())
+					return
+				}
+
+				modifiedPod := pod.DeepCopy()
+				if !modifyPodSpec(modifiedPod) {
+					log.Printf("no injection in pod/%s", pod.GetName())
+				}
+
+				removeSelfPendingInitializer(modifiedPod)
+
+				if err := patchPod(pod, modifiedPod, clientset); err != nil {
+					log.Printf("error saving pod/%s: %+v", pod.GetName(), err)
+				} else {
+					log.Printf("initialized pod/%s", pod.GetName())
+				}
+			},
+		},
+	)
+
+	stop := make(chan struct{})
+	go controller.Run(stop)
+
+	signalChan := make(chan os.Signal, 1)
+	signal.Notify(signalChan, syscall.SIGINT, syscall.SIGTERM)
+	<-signalChan
+
+	log.Println("Shutdown signal received, exiting...")
+	close(stop)
+}
+
+// needsInitialization determines if the pod is required to be initialized
+// currently by this initializer.
+func needsInitialization(pod *corev1.Pod) bool {
+	initializers := pod.ObjectMeta.GetInitializers()
+	return initializers != nil &&
+		len(initializers.Pending) > 0 &&
+		initializers.Pending[0].Name == initializerName
+}
+
+// removeSelfPendingInitializer removes the first element from pending
+// initializers list of in-memory pod value.
+func removeSelfPendingInitializer(pod *corev1.Pod) {
+	if pod.ObjectMeta.GetInitializers() == nil {
+		return
+	}
+	pendingInitializers := pod.ObjectMeta.GetInitializers().Pending
+	if len(pendingInitializers) == 1 {
+		pod.ObjectMeta.Initializers.Pending = nil
```

Comment:
I don't have the exact error message at hand, but this is how I make it work for me. 
pod.ObjectMeta.Initializers = nil

Reply:
hmm it would be good to repro this. it actually works like this for example/nginx.yaml

Reply:
@fenglu-g the error you saw was fixed in 1.8.